reference-delegate-explainer.md

```diff
@@ -0,0 +1,534 @@
+- The solution should be serializable, i.e. support declarative syntax that is expressible in HTML only.
+
+**Non-Goals**
+
+- This is scoped to only solve the problem of referring _into_ the shadow DOM. It relies on ARIAMixin to refer _out_ of the shadow DOM.
+- This proposal does not delegate ARIA attributes that aren't ID references, such as `aria-label`.
+
+### The `referenceDelegate` attribute
+
+A component can specify an element in its shadow tree to act as its "reference delegate". When the host component is the target of a IDREF like a label's `for` attribute, the delegate becomes the true target.
+
+The shadow root specifies the ID of the delegate element inside the shadow DOM. This is done either in JavaScript with the `referenceDelegate` attribute on the `ShadowRoot` object, or in HTML markup using the `shadowrootreferencedelegate` attribute on the `<template>` element.
+
+JavaScript example:
+
+```html
+<template id="t-fancy-input">
+  <input id="real-input" />
+</template>
+
+<script>
+customElements.define("fancy-input", 
+  class FancyInput extends HTMLElement {
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.appendChild(
+        document.getElementById("t-fancy-input").content.cloneNode(true));
+      this.shadowRoot_.referenceDelegate = "real-input";
+    }
+  });
+</script>
+
+<label for="fancy-input">Fancy input</label>
+<fancy-input id="fancy-input"></fancy-input>
+```
+
+Equivalent with declarative shadow DOM:
+
+```html
+<label for="fancy-input">Fancy input</label>
+<fancy-input id="fancy-input">
+  <template shadowrootmode="closed"
+            shadowrootreferencedelegate="real-input">
+    <input id="real-input" />
+  </template>
+</fancy-input>
+```
+
+### The `referenceDelegateMap` attribute
+
+There are situations where it is necessary to delegate different reference types to different elements. For example, a listbox may want to delegate `aria-controls` to its root, and `aria-activedescendant` to one of the items inside the listbox.
+
+The `ShadowRoot.referenceDelegateMap` attribute allows for specifying elements based on the attribute that is being used to reference the host.
+
+The equivalent declarative attribute is `shadowrootreferencedelegatemap`, which is a comma-separated list of attribute to ID mappings.
+
+```html
+<input role="combobox"
+       aria-controls="fancy-listbox"
+       aria-activedescendant="fancy-listbox" />
+<fancy-listbox id="fancy-listbox">
+  <template shadowrootmode="closed"
+            shadowrootreferencedelegatemap="aria-controls: real-listbox,
+                                            aria-activedescendant: option-2">
+    <div id="real-listbox" role="listbox">
+      <div id="option-1" role="option">Option 1</div>
+      <div id="option-2" role="option">Option 2</div>
+    </div>
+  </template>
+</fancy-listbox>
+```
```

Comment:
Does this approach allow a user to do the following?
```html
<fancy-input
  role="combobox"
  aria-controls="fancy-listbox"
  aria-activedescendant="fancy-listbox"
>
  <template shaodwrootmode="open" shadowrootreferencedelegate="input">
    <input id="input" />
  </template>
</fancy-input>
<fancy-listbox id="fancy-listbox">
  <template shadowrootmode="open"
            shadowrootreferencedelegatemap="aria-controls: real-listbox,
                                            aria-activedescendant: option-2">
    <div id="real-listbox" role="listbox">
      <div id="option-1" role="option">Option 1</div>
      <div id="option-2" role="option">Option 2</div>
    </div>
  </template>
</fancy-listbox>
```
There's historically been a lot of talk about needing a holistic solution to clarify the path after a "good enough" solution, and I'm not sure what that would be in this case.

Reply:
To clarify: this is asking whether the `role=combobox` is projected down on to the `<input>`?